File: opencode-plugin/tests/mcp_sdk_interop.ts
/**
 * Driver for `tests/mcp_sdk_interop.py` — it runs the plugin's real hooks against a REAL
 * `mcp`-SDK server (fixture tools, no Postgres), started by the Python side.
 *
 * Not runnable on its own: the runner sets SCAFFOLD_ENGINE_URL and INTEROP_FIXTURES, then
 * reads the `__RESULT__` line this driver prints. Assertions about what actually arrived at
 * the server (and that no source text ever did) live on the Python side.
 */

const fixtures = JSON.parse(process.env.INTEROP_FIXTURES ?? "{}")
const engineUrl = process.env.SCAFFOLD_ENGINE_URL ?? ""
if (!engineUrl) {
  console.error("SCAFFOLD_ENGINE_URL is required (set by mcp_sdk_interop.py)")
  process.exit(2)
}

const { ScaffoldPlugin } = await import(new URL("../.opencode/plugins/scaffold.ts", import.meta.url).href)

const checks: Array<{ name: string; ok: boolean; detail?: string }> = []
const check = (name: string, cond: unknown, detail = "") =>
  checks.push({ name, ok: cond === true, detail: cond === true ? "" : detail })

const logs: any[] = []
const toasts: any[] = []
const client = {
  app: { log: async ({ body }: any) => void logs.push(body) },
  tui: { showToast: async ({ body }: any) => void toasts.push(body) },
}
const WORKTREE = process.cwd()
const hooks = await ScaffoldPlugin({
  client,
  project: { id: "opencode-project-hash" },
  directory: WORKTREE,
  worktree: WORKTREE,
  $: async () => {},
} as any)

/** Only what the plugin appended: `system[length-1]` would be the base prompt when nothing
 *  was injected, which turns "contains X" checks into silent no-ops. */
const systemBlock = async () => {
  const system: string[] = ["base system prompt"]
  const baseEntries = system.length
  await hooks["experimental.chat.system.transform"]!({ sessionID: "ses-interop" }, { system } as any)
  const appended = system.slice(baseEntries)
  return { block: appended.join("\n\n"), appended: appended.length }
}

// 1. session start → context pulled over the real protocol (initialize + session id + SSE)
await hooks.event!({ event: { type: "session.created", properties: { sessionID: "ses-interop" } } } as any)
const first = await systemBlock()
check(
  "context block injected from the real MCP server",
  first.appended === 1 && first.block.includes(fixtures.projectName),
  JSON.stringify(first.block.slice(0, 120)),
)
check("block carries the fixture decision", first.block.includes(fixtures.decisionText))
check("session start toasted the loaded context", toasts.some((t) => /context loaded/.test(t.message)))

// 2. a user message that names the route, then a write to it → get_api_contract(route)
await hooks["chat.message"]!({ sessionID: "ses-interop" } as any)
const SOURCE = fixtures.sourceMarker
await hooks["tool.execute.before"]!(
  { tool: "write", sessionID: "ses-interop", callID: "call-1" },
  {
    args: {
      filePath: `${WORKTREE}/engine/app/routes/interop.py`,
      content: `@router.post("${fixtures.route}")\ndef interop():\n    # ${SOURCE}\n    pass\n`,
    },
  },
)
const pinned = await systemBlock()
check(
  "pinned contract reaches the next request",
  pinned.appended === 1 && pinned.block.includes(`${fixtures.method} ${fixtures.route}`),
  pinned.block.slice(-200),
)
check("pinned contract carries the registered request shape", pinned.block.includes(fixtures.schemaMarker))

// 3. a real edit → report_change over the real protocol
await hooks["tool.execute.after"]!(
  { tool: "edit", sessionID: "ses-interop", callID: "call-2", args: { filePath: `${WORKTREE}/engine/app/routes/interop.py` } },
  {
    title: "engine/app/routes/interop.py",
    output: "Edit applied successfully.",
    metadata: { filediff: { file: `${WORKTREE}/engine/app/routes/interop.py`, additions: 7, deletions: 2 } },
  } as any,
)

// 4. shell env
const env = { env: {} as Record<string, string> }
await hooks["shell.env"]!({ cwd: WORKTREE }, env as any)
check("shell.env still injects the frozen var", env.env.SCAFFOLD_ENGINE_URL === engineUrl)

check("no source text appears in plugin logs", !JSON.stringify(logs).includes(SOURCE))
check("no warnings while the engine is healthy", logs.every((l) => l.level !== "warn"), JSON.stringify(logs.filter((l) => l.level === "warn")))

console.log(`__RESULT__${JSON.stringify({ checks, logged: logs.length, toasts: toasts.length })}`)
// Undici keeps sockets alive; the runner needs a clean exit.
process.exit(checks.every((c) => c.ok) ? 0 : 1)

File: opencode-plugin/tests/mcp_sdk_interop.py
"""Day 4 interop check: the plugin's hand-rolled MCP client vs the REAL `mcp` SDK server.

Why this exists: `tests/verify_plugin.ts` proves the plugin's hook logic, but it drives a fake
engine written by the same hand that wrote the client — so it cannot catch a wire mismatch.
`engine/tests/test_day2.py` proves the engine's *server* speaks this protocol to a *Python*
client. This closes the loop: a real `mcp`-SDK server (the fixture in `engine_fixture.py`, built
with the engine's own mount code) with fixture tools — no Postgres, no engine app, no LLM.

Run it from anywhere (it re-executes itself with engine/.venv's interpreter if needed):

    python opencode-plugin/tests/mcp_sdk_interop.py
"""

import json
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from engine_fixture import FIXTURES, FixtureEngine, check, report, run_driver  # noqa: E402

DRIVER = Path(__file__).resolve().with_name("mcp_sdk_interop.ts")

with FixtureEngine() as engine:
    print(f"\n== real mcp SDK server on {engine.url}/mcp ==")
    driver, proc = run_driver(DRIVER, engine.url)

    if driver is None:
        print(proc.stdout)
        print(proc.stderr)
        print("FAIL  the plugin driver produced no result (see output above)")
        sys.exit(1)

    for entry in driver["checks"]:
        check(entry["name"], entry["ok"], entry.get("detail", ""))
    check("plugin driver exited cleanly", proc.returncode == 0, f"exit {proc.returncode}")

    called = [call["tool"] for call in engine.calls]
    check(
        "the real SDK server received all three frozen tool names",
        set(called) == {"get_project_context", "get_api_contract", "report_change"},
        str(called),
    )
    check(
        "context is refetched exactly once per user message (cache invalidation over the real protocol)",
        called.count("get_project_context") == 2 and engine.calls[0]["tool"] == "get_project_context",
        str(called),
    )
    check(
        "the plugin sent no project_id (engine-side SCAFFOLD_DEFAULT_PROJECT_ID convention)",
        all(call["project_id"] is None for call in engine.calls),
        str(engine.calls),
    )
    contract_call = next((call for call in engine.calls if call["tool"] == "get_api_contract"), {})
    check(
        "contract lookup used the route from the write arguments",
        contract_call.get("route") == FIXTURES["route"],
        str(contract_call),
    )
    check("exactly one change was reported", len(engine.reports) == 1, str(len(engine.reports)))

    sent = engine.reports[0] if engine.reports else {}
    check(
        "reported summary is the deterministic diff metadata",
        sent.get("diff_summary") == "edited engine/app/routes/interop.py (+7 -2 lines)",
        str(sent),
    )
    check("reported files are worktree-relative", sent.get("files_changed") == ["engine/app/routes/interop.py"], str(sent))
    check(
        "no source text crossed the wire (repo rule #4)",
        FIXTURES["sourceMarker"] not in json.dumps({"calls": engine.calls, "reports": engine.reports}),
    )

    if driver["checks"] and not all(c["ok"] for c in driver["checks"]):
        print(proc.stdout)
        print(proc.stderr)

sys.exit(report())
